source/extensions/transport_sockets/tls/utility.cc

```diff
@@ -119,6 +119,40 @@ int32_t Utility::getDaysUntilExpiration(const X509* cert, TimeSource& time_sourc
   return 0;
 }
 
+absl::optional<std::string> Utility::getX509ExtensionValue(const X509& cert,
+                                                           absl::string_view extension_name) {
+  X509_EXTENSIONS* extensions(X509_get0_extensions(&cert));
+
+  if (extensions == nullptr) {
+    return absl::nullopt;
+  }
+
+  const size_t extension_count = sk_X509_EXTENSION_num(extensions);
+
+  for (size_t i = 0; i < extension_count; ++i) {
+    X509_EXTENSION* extension = sk_X509_EXTENSION_value(extensions, i);
+
+    ASN1_OBJECT* extension_object = X509_EXTENSION_get_object(extension);
+    const size_t size = OBJ_obj2txt(nullptr, 0, extension_object, 0);
+    std::vector<char> buffer;
+    // +1 to allow for NULL byte.
+    buffer.resize(size + 1);
+    OBJ_obj2txt(buffer.data(), buffer.size(), extension_object, 0);
+
+    if (absl::string_view(buffer.data(), size) == extension_name) {
+      ASN1_OCTET_STRING* octet_string = X509_EXTENSION_get_data(extension);
```

Comment:
QQ: If envoy get a malformed extension, where the value of the extension is not an ASN_OCTET_STRING type, would envoy be failing elsewhere? Below code blindly extract the string anyway...

Reply:
I would suspect that if the cert has invalid extensions it would fail earlier than this, but not sure? @PiotrSikora @lizan any idea?

Reply:
I'm curious about this also. The API isn't documented very well, so I can't tell what would happen. As an easy test, can you convince openssl to generate a cert with a different type of extension?

Reply:
I added a test case where I just add a DER encoded byte sequence, definitely doesn't crash (though I don't know enough about these formats to understand if the resulting output is expected)

I guess an alternative approach is to verify that the tag is OCTET_STRING and restrict this function to only work for that case?

Reply:
The docs I could find from openssl said that these extensions should all be of type OCTET_STRING, so if you validated that it doesn't crash, I think it's probably fine. Up to you if you want to add an additional check.